File: apps/ios/Wordocious/Sources/TodaysRace.swift
import SwiftUI
import WordociousCore

/// TODAY'S RACE (Friends D3.1, §289) — the friend leaderboard for TODAY: you
/// and every friend ranked by today's daily points (the digest's todayPoints:
/// composite score summed over every daily played today, sweep and More Games
/// alike). Ties share a rank (competition ranking), then sort alphabetically.
/// Pure twin of apps/web/lib/todays-race.ts so the three platforms agree.
enum TodaysRace {
    struct Entrant {
        let id: String
        let username: String
        let points: Int
        /// Sweep dailies played today (the "N/8" line).
        let played: Int
        let me: Bool
    }

    struct Row: Identifiable {
        let id: String
        let username: String
        let points: Int
        let played: Int
        let me: Bool
        let rank: Int
    }

    static func rankToday(_ entrants: [Entrant]) -> [Row] {
        let sorted = entrants.sorted { a, b in
            if a.points != b.points { return a.points > b.points }
            return a.username.localizedCompare(b.username) == .orderedAscending
        }
        var out: [Row] = []
        for (i, e) in sorted.enumerated() {
            let rank = (out.last.map { $0.points == e.points } ?? false) ? out.last!.rank : i + 1
            out.append(Row(id: e.id, username: e.username, points: e.points, played: e.played, me: e.me, rank: rank))
        }
        return out
    }

    /// "Leading by 340" / "120 behind Doug" / "Tied with Doug" / "Play a daily to join the race".
    static func raceStatusLine(_ rows: [Row]) -> String {
        guard let me = rows.first(where: { $0.me }) else { return "" }
        if me.points == 0 { return "Play a daily to join the race" }
        let others = rows.filter { !$0.me }
        guard let next = others.first else { return "\(me.points.formatted()) pts today" }
        if me.rank == 1 {
            return next.points == me.points ? "Tied with \(next.username)" : "Leading by \((me.points - next.points).formatted())"
        }
        let ahead = others.filter { $0.points > me.points }.min { $0.points < $1.points }!
        return "\((ahead.points - me.points).formatted()) behind \(ahead.username)"
    }
}

/// The TODAY'S RACE section at the top of the FRIENDS card (§289; twin of
/// components/friends/todays-race.tsx). Every row has a reason to tap:
/// Challenge (a private Classic VS Battle, pushed to the friend, free for
/// friends) and the bell for a friend who hasn't played yet. Tapping a
/// friend's row opens their profile — the same NavigationLink(value:) push
/// the roster rows use; your own row goes nowhere.
struct TodaysRaceCard: View {
    let friends: [FriendsService.FriendProfile]
    let me: Profile
    let meDigest: FriendsService.MeDigest?
    /// The friend id whose challenge is in flight (dims the other buttons).
    let challenging: String?
    let onTaunt: (FriendsService.FriendProfile) -> Void
    let onChallenge: (FriendsService.FriendProfile) -> Void

    private static let purple = Color(hex: 0x7C3AED)
    private static let pink = Color(hex: 0xEC4899)

    var body: some View {
        let rows = TodaysRace.rankToday(
            friends.map { .init(id: $0.id, username: $0.username, points: $0.todayPoints ?? 0, played: $0.playedToday ?? 0, me: false) }
            + [.init(id: me.id, username: me.username, points: meDigest?.todayPoints ?? 0, played: meDigest?.playedToday ?? 0, me: true)])
        let byId = Dictionary(uniqueKeysWithValues: friends.map { ($0.id, $0) })
        let anyPoints = rows.contains { $0.points > 0 }

        VStack(alignment: .leading, spacing: 6) {
            HStack(spacing: 5) {
                Image(systemName: "flag.fill").font(.system(size: 11, weight: .bold))
                    .foregroundStyle(Self.purple)
                Text("TODAY'S RACE").font(Brand.font(9, .black)).tracking(0.8)
                    .foregroundStyle(Theme.textMuted)
                Spacer(minLength: 6)
                Text(TodaysRace.raceStatusLine(rows))
                    .font(Brand.font(10, .black))
                    .foregroundStyle(anyPoints ? Self.purple : Theme.textMuted)
                    .lineLimit(1).minimumScaleFactor(0.8)
            }
            VStack(spacing: 6) {
                ForEach(rows) { r in
                    if r.me {
                        row(r, friend: nil, anyPoints: anyPoints)
                    } else if let f = byId[r.id] {
                        NavigationLink(value: r.id) {
                            row(r, friend: f, anyPoints: anyPoints)
                        }
                        .buttonStyle(.plain)
                    }
                }
            }
            Text("Today's points across every daily · Challenge = a private Classic battle, free for friends")
                .font(Brand.font(9, .bold)).foregroundStyle(Theme.textMuted)
                .frame(maxWidth: .infinity)
                .multilineTextAlignment(.center)
        }
        .padding(.vertical, 4)
    }

    @ViewBuilder
    private func row(_ r: TodaysRace.Row, friend f: FriendsService.FriendProfile?, anyPoints: Bool) -> some View {
        let medal: String? = anyPoints && r.points > 0 && r.rank <= 3 ? ["🥇", "🥈", "🥉"][r.rank - 1] : nil
        HStack(spacing: 10) {
            Text(medal ?? "\(r.rank)")
                .font(medal == nil ? Brand.font(11, .black) : .system(size: 14))
                .foregroundStyle(Theme.textMuted)
                .frame(width: 24)
            if let f {
                AvatarView(url: f.avatar_url, username: f.username, size: 30, emoji: f.avatar_emoji)
            } else {
                AvatarView(url: me.avatarUrl, username: me.username, size: 30, emoji: me.avatarEmoji)
            }
            VStack(alignment: .leading, spacing: 1) {
                Text(r.me ? "You" : r.username).font(Brand.font(12, .heavy))
                    .foregroundStyle(r.me ? Self.purple : Theme.textPrimary).lineLimit(1)
                Text(r.points > 0
                     ? "\(r.points.formatted()) pts · \(r.played)/\(DailyCompletionsStore.totalDailyModes) dailies"
                     : "hasn't played today")
                    .font(Brand.font(10, .bold)).foregroundStyle(Theme.textMuted).lineLimit(1)
            }
            Spacer(minLength: 4)
            if let f {
                if r.points == 0 {
                    // Slacker bell — the existing canned-taunt picker.
                    Button { onTaunt(f) } label: {
                        Image(systemName: "bell.fill").font(.system(size: 12, weight: .bold))
                            .foregroundStyle(Self.purple)
                            .frame(width: 26, height: 26)
                            .background(Circle().fill(Theme.surfaceAlt))
                            .overlay(Circle().stroke(Theme.border, lineWidth: 1.5))
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel("Nudge \(f.username)")
                }
                Button { onChallenge(f) } label: {
                    HStack(spacing: 4) {
                        Image("swords").renderingMode(.template).resizable().scaledToFit()
                            .frame(width: 11, height: 11)
                        Text(challenging == f.id ? "Sending…" : "Challenge").font(Brand.font(10, .black))
                    }
                    .foregroundStyle(Self.pink)
                    .padding(.horizontal, 8).padding(.vertical, 5)
                    .background(RoundedRectangle(cornerRadius: 8).fill(Self.pink.opacity(0.08)))
                    .overlay(RoundedRectangle(cornerRadius: 8).stroke(Self.pink, lineWidth: 1.5))
                }
                .buttonStyle(.plain)
                .disabled(challenging != nil)
                .opacity(challenging != nil && challenging != f.id ? 0.5 : 1)
                .accessibilityLabel("Challenge \(f.username) to a VS Battle")
            }
        }
        .padding(.horizontal, 8).padding(.vertical, 6)
        .background(RoundedRectangle(cornerRadius: 12).fill(r.me ? Self.purple.opacity(0.06) : Color.clear))
        .overlay(RoundedRectangle(cornerRadius: 12).stroke(r.me ? Color(hex: 0xC4B5FD) : Color.clear, lineWidth: 1))
    }
}

File: apps/android/app/src/main/kotlin/com/wordocious/app/ui/TodaysRace.kt
package com.wordocious.app.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Flag
import androidx.compose.material.icons.outlined.Notifications
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.wordocious.app.ModeGen
import com.wordocious.app.R
import com.wordocious.app.data.FriendsService
import com.wordocious.app.ui.theme.Nunito
import com.wordocious.app.ui.theme.WTheme

// TODAY'S RACE (§289, Friends D3.1) — the top of the FRIENDS card: you and
// every friend ranked by TODAY's daily points (the digest's todayPoints —
// composite score over every daily played today, sweep and More Games alike).
// Ties share a rank (competition ranking), then alphabetical. Every row has a
// reason to tap: Challenge (a private Classic VS Battle, pushed to the friend,
// free for friends), the bell for a friend who hasn't played, the row itself
// into the profile. Twin of web components/friends/todays-race.tsx; the
// ranking below mirrors lib/todays-race.ts and is pinned by TodaysRaceTest.

data class RaceEntrant(
    val id: String,
    val username: String,
    val points: Int,
    /** Sweep dailies played today (the "N/8" line). */
    val played: Int,
    val me: Boolean,
)

data class RaceRow(
    val id: String,
    val username: String,
    val points: Int,
    val played: Int,
    val me: Boolean,
    val rank: Int,
)

/** Points desc, ties share a rank, then alphabetical — lib/todays-race.ts rankToday. */
fun rankToday(entrants: List<RaceEntrant>): List<RaceRow> {
    val sorted = entrants.sortedWith(
        compareByDescending<RaceEntrant> { it.points }
            .thenBy(String.CASE_INSENSITIVE_ORDER) { it.username }
            .thenBy { it.username },
    )
    val out = ArrayList<RaceRow>(sorted.size)
    sorted.forEachIndexed { i, e ->
        val prev = out.getOrNull(i - 1)
        val rank = if (prev != null && prev.points == e.points) prev.rank else i + 1
        out.add(RaceRow(e.id, e.username, e.points, e.played, e.me, rank))
    }
    return out
}

/** "Leading by 340" / "120 behind Doug" / "Tied with Doug" / "Play a daily to join the race". */
fun raceStatusLine(rows: List<RaceRow>): String {
    val me = rows.firstOrNull { it.me } ?: return ""
    if (me.points == 0) return "Play a daily to join the race"
    val others = rows.filter { !it.me }
    if (others.isEmpty()) return "${racePts(me.points)} pts today"
    if (me.rank == 1) {
        val next = others.first()
        return if (next.points == me.points) "Tied with ${next.username}"
        else "Leading by ${racePts(me.points - next.points)}"
    }
    val ahead = others.filter { it.points > me.points }.minByOrNull { it.points }!!
    return "${racePts(ahead.points - me.points)} behind ${ahead.username}"
}

/** "1,504" — US-grouped points, the race surfaces' formatting. */
internal fun racePts(n: Int): String = String.format(java.util.Locale.US, "%,d", n)

private val PURPLE = Color(0xFF7C3AED)
private val PINK = Color(0xFFEC4899)

/**
 * The card body. Renders nothing without a signed-in profile or friends.
 * [challengingId] is the friend whose challenge is in flight (one at a time).
 */
@Composable
fun TodaysRaceCard(
    friends: List<FriendsService.FriendProfile>,
    meDigest: FriendsService.MeDigest?,
    challengingId: String?,
    onOpenProfile: (String) -> Unit,
    onTaunt: (FriendsService.FriendProfile) -> Unit,
    onChallenge: (FriendsService.FriendProfile) -> Unit,
) {
    val me = com.wordocious.app.data.AuthService.profile.value ?: return
    if (friends.isEmpty()) return

    val rows = rankToday(
        friends.map { RaceEntrant(it.id, it.username, it.todayPoints ?: 0, it.playedToday ?: 0, me = false) } +
            RaceEntrant(me.id, me.username ?: "You", meDigest?.todayPoints ?: 0, meDigest?.playedToday ?: 0, me = true),
    )
    val byId = friends.associateBy { it.id }
    val status = raceStatusLine(rows)
    val anyPoints = rows.any { it.points > 0 }
    val sweepSize = ModeGen.sweep.size

    Column(verticalArrangement = Arrangement.spacedBy(6.dp), modifier = Modifier.fillMaxWidth()) {
        Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.fillMaxWidth()) {
            Icon(Icons.Filled.Flag, null, tint = PURPLE, modifier = Modifier.size(13.dp))
            Spacer(Modifier.width(4.dp))
            Text(
                "TODAY'S RACE", fontSize = 9.sp, fontWeight = FontWeight.Black,
                letterSpacing = 0.8.sp, color = WTheme.textMuted, fontFamily = Nunito,
            )
            Spacer(Modifier.weight(1f))
            Text(
                status, fontSize = 10.sp, fontWeight = FontWeight.Black,
                color = if (anyPoints) PURPLE else WTheme.textMuted, fontFamily = Nunito,
                maxLines = 1, overflow = TextOverflow.Ellipsis,
            )
        }
        rows.forEach { r ->
            val f = if (r.me) null else byId[r.id]
            val medal = if (anyPoints && r.points > 0) listOf("🥇", "🥈", "🥉").getOrNull(r.rank - 1) else null
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                modifier = Modifier
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(12.dp))
                    .background(if (r.me) PURPLE.copy(alpha = 0.06f) else Color.Transparent)
                    .border(1.dp, if (r.me) Color(0xFFC4B5FD) else Color.Transparent, RoundedCornerShape(12.dp))
                    .padding(horizontal = 6.dp, vertical = 5.dp),
            ) {
                Text(
                    medal ?: "${r.rank}", fontSize = 11.sp, fontWeight = FontWeight.Black,
                    color = WTheme.textMuted, fontFamily = Nunito,
                    textAlign = TextAlign.Center, modifier = Modifier.width(22.dp),
                )
                // The row opens the friend's profile; your own row goes nowhere.
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                    modifier = Modifier.weight(1f).let { m ->
                        if (f != null) m.clickableNoRipple { onOpenProfile(f.id) } else m
                    },
                ) {
                    FriendAvatar(
                        f ?: FriendsService.FriendProfile(
                            id = me.id, username = me.username ?: "You",
                            avatarUrl = me.avatarUrl, avatarEmoji = me.avatarEmoji, level = me.level,
                        ),
                    )
                    Column(verticalArrangement = Arrangement.spacedBy(1.dp), modifier = Modifier.weight(1f)) {
                        Text(
                            if (r.me) "You" else r.username, fontSize = 12.sp,
                            fontWeight = FontWeight.ExtraBold,
                            color = if (r.me) PURPLE else WTheme.text,
                            maxLines = 1, overflow = TextOverflow.Ellipsis,
                        )
                        Text(
                            if (r.points > 0) "${racePts(r.points)} pts · ${r.played}/$sweepSize dailies"
                            else "hasn't played today",
                            fontSize = 10.sp, fontWeight = FontWeight.Bold,
                            color = WTheme.textMuted, fontFamily = Nunito,
                            maxLines = 1, overflow = TextOverflow.Ellipsis,
                        )
                    }
                }
                if (f != null) {
                    if (r.points == 0) {
                        // The bell — the existing canned taunt picker.
                        Box(
                            Modifier.size(26.dp).clip(CircleShape).background(WTheme.surfaceHover)
                                .border(1.5.dp, WTheme.border, CircleShape)
                                .clickableNoRipple { onTaunt(f) },
                            contentAlignment = Alignment.Center,
                        ) {
                            Icon(
                                Icons.Outlined.Notifications, "Nudge ${f.username}",
                                tint = PURPLE, modifier = Modifier.size(14.dp),
                            )
                        }
                    }
                    val sending = challengingId == f.id
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(4.dp),
                        modifier = Modifier
                            .clip(RoundedCornerShape(8.dp))
                            .background(PINK.copy(alpha = 0.08f))
                            .border(1.5.dp, PINK, RoundedCornerShape(8.dp))
                            .alpha(if (challengingId != null && !sending) 0.5f else 1f)
                            .clickableNoRipple { if (challengingId == null) onChallenge(f) }
                            .padding(horizontal = 8.dp, vertical = 4.dp),
                    ) {
                        Icon(
                            painterResource(R.drawable.ic_swords), "Challenge ${f.username} to a VS Battle",
                            tint = PINK, modifier = Modifier.size(11.dp),
                        )
                        Text(
                            if (sending) "Sending…" else "Challenge", fontSize = 10.sp,
                            fontWeight = FontWeight.Black, color = PINK, fontFamily = Nunito,
                        )
                    }
                }
            }
        }
        Text(
            "Today's points across every daily · Challenge = a private Classic battle, free for friends",
            fontSize = 9.sp, fontWeight = FontWeight.Bold, color = WTheme.textMuted, fontFamily = Nunito,
            textAlign = TextAlign.Center, modifier = Modifier.fillMaxWidth().padding(top = 2.dp),
        )
    }
}
